packages/cohort/src/tx_batch_executor.rs

```diff
@@ -24,97 +24,90 @@ impl BatchExecutor {
-
         if action_result.is_err() {
-            let action_error = action_result.unwrap_err();
-            let tx_error = tx.rollback().await.map_err(|e| e.to_string());
+            return Err(Self::handle_rollback(tx.rollback().await, action_result.unwrap_err()));
+        }
 
-            if tx_error.is_err() {
-                Err(format!(
-                    "Cannot rollback failed action. Error: {:?}, Rollback error: {}",
-                    action_error,
-                    tx_error.unwrap_err(),
-                ))
-            } else {
-                Err(format!("Cannot execute batch. Error: {:?}. Rollback.", action_error))
-            }
-        } else {
-            let mut affected_rows = 0_u64;
-            for c in action_result.unwrap().iter() {
-                affected_rows += c;
-            }
+        let mut affected_rows = 0_u64;
+        // filter out empty Option elements, here flatten() = filter(Option::is_some).map(Option::unwrap)
+        for c in action_result.unwrap().iter().flatten() {
+            affected_rows += c;
+        }
 
-            if affected_rows == 0 {
-                // still fail here
-                let tx_error = tx.rollback().await.map_err(|e| e.to_string());
-                if tx_error.is_err() {
-                    Err(format!(
-                        "Cannot rollback failed action. Error: No rows where updated. Rollback error: {}",
-                        tx_error.unwrap_err(),
-                    ))
-                } else {
-                    Err("Cannot execute action. Error: No rows where updated".to_string())
-                }
-            } else {
-                //
-                if let Some(new_version) = snapshot {
-                    let snapshot_update_result = SnapshotApi::update_using(&tx, new_version).await;
-                    if let Ok(rows) = snapshot_update_result {
-                        affected_rows += rows;
-                    } else {
-                        let snapshot_error = snapshot_update_result.unwrap_err();
-                        // there was an error updating snapshot, we need to rollabck the whole batch
-                        let tx_error = tx.rollback().await.map_err(|e| e.to_string());
-                        return if tx_error.is_err() {
-                            Err(format!(
-                                "Cannot rollback after snapshot failed to update. Error: {}. Rollback error: {}",
-                                snapshot_error,
-                                tx_error.unwrap_err(),
-                            ))
-                        } else {
-                            Err(format!("Cannot update snapshot. Error: {}. Rollback", snapshot_error))
-                        };
-                    }
-                }
+        if affected_rows == 0 {
+            // still fail here
+            return Err(Self::handle_rollback(tx.rollback().await, "No rows where updated".to_string()));
```

Comment:
I feel this shouldn't throw an error, but an early return with 0 rows. And maybe print a warning?

Reply:
Because sometimes, replicator has already moved ahead and if initiator is already installed it, we wont install it again. 

And vice-versa.